File: Daily Resources/Day05/DSA - Prefix and Difference Arrays Solutions/divisible_by_k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        unordered_map<int, int> count;
        count[0] = 1;

        int prefix_sum = 0;
        int result = 0;

        for (int num : nums) {
            prefix_sum += num;
            int mod = ((prefix_sum % k) + k) % k; // handle negative mods

            result += count[mod];
            count[mod]++;
        }

        return result;
    }
};

File: Daily Resources/Day05/DSA - Prefix and Difference Arrays Solutions/divisible_by_k.py
from collections import defaultdict

class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        count = defaultdict(int)
        count[0] = 1  # one empty prefix with mod 0

        prefix_sum = 0
        result = 0

        for num in nums:
            prefix_sum += num
            mod = prefix_sum % k
            if mod < 0:
                mod += k  # ensure non-negative remainder

            result += count[mod]
            count[mod] += 1

        return result
